util/network-alert/src/notifier.rs

```diff
@@ -0,0 +1,90 @@
+use ckb_core::alert::Alert;
+use fnv::FnvHashMap;
+use log::warn;
+use lru_cache::LruCache;
+use std::sync::Arc;
+
+const CANCEL_FILTER_SIZE: usize = 128;
+
+pub struct Notifier {
+    /// cancelled alerts
+    cancel_filter: LruCache<u32, ()>,
+    /// alerts we received
+    received_alerts: FnvHashMap<u32, Arc<Alert>>,
+    /// alerts that self node should notice
+    noticed_alerts: Vec<Arc<Alert>>,
+    client_version: String,
+}
+
+impl Notifier {
+    pub fn new(client_version: String) -> Self {
+        Notifier {
+            cancel_filter: LruCache::new(CANCEL_FILTER_SIZE),
+            received_alerts: Default::default(),
+            noticed_alerts: Vec::new(),
+            client_version,
+        }
+    }
+
+    pub fn add(&mut self, alert: Arc<Alert>) {
+        if self.has_received(alert.id) {
+            return;
+        }
+        // checkout cancel_id
+        if alert.cancel > 0 {
+            self.cancel(alert.cancel);
+        }
+        // add to received alerts
+        self.received_alerts.insert(alert.id, Arc::clone(&alert));
```

Comment:
`if insert.is_some() return;` ?

Reply:
Can we merge `noticed_alerts` and `received_alerts`?

Reply:
There are for the different purpose, `noticed_alerts` store alerts that the current node should be noticed through RPC or other approaches. 

`received_alerts` store all alerts we received, even the current node ignores, we still store them for relay to other peers.

Reply:
Got it.

So what if we give the alert messages some identifiers, like

```
enum {
  Cancel(alert),
  Normal(alert),
  ...
}
```

I'm not sure which one is simpler.